torchtitan/datasets/experimental_datasets.py

```diff
@@ -0,0 +1,1212 @@
+    end = math.ceil(len(itemlist) * (rank + 1) / worldsize)
+    return itemlist[start:end]
+
+
+class _Stateful_Dataset(data.IterableDataset):
+    """
+    Stub for stateful datasets, extends data.IterableDataset with state_dict methods.
+    All subclasses should specify the params to be considered stateful or reshardable in the
+    self.state_params and self.reshard_params lists.
+    """
+
+    def __init__(
+        self,
+        rank: int,
+        worldsize: int,
+    ):
+        assert rank >= 0, f"Rank {rank} must be a positive integer"
+        assert (
+            worldsize > rank
+        ), f"Worldsize {worldsize} must be greater than rank {rank}"
+        self.state_params: List[str] = []
+        self.reshard_params: List[str] = []
+        self.rank = rank
+        self.worldsize = worldsize
+        self.load_worldsize = (
+            worldsize  # Enable calling load_state_dict() directly, assume no rescaling
+        )
+
+    def statename(self, x: str):
+        # Note that this naming convention implicitly disallows repeated layers in the dataset pipeline
+        return self.__class__.__name__ + "." + x
+
+    def state_dict(self):
+        """
+        Retrieve all state and reshard flags (each worker/process saves its own state dict shard)
+        """
+        return {
+            self.statename(flag): getattr(self, flag)
+            for flag in self.state_params + self.reshard_params
+        }
+
+    def _reshard(self, sharded_list):
+        """
+        Sharded_list is a list of lists, where each "shard" sublist must have the same length.
+        These shards should tightly span only the partition of data owned by this worker.
+        (i.e. if global_list is the list of all entries, sharded_list = _shard_inclusive(global_list) ).
+        Determine fractional ownership of shards, and get the flattened partition owned by this worker.
+        """
+        # How many shards did _shard_inclusive() drop to the left of sharded_list?
+        shard_offset = math.floor(self.load_worldsize * self.rank / self.worldsize)
+        # How long are the list shards?
+        shard_len = len(sharded_list[0])
+        for i, shard in enumerate(sharded_list):
+            assert (
+                len(shard) == shard_len
+            ), f"Shard {i} with length {len(shard)} does not match expected {shard_len}"
+        # How many list items did _shard_inclusive() drop to the left of the flattened sharded_list?
+        item_offset = shard_len * shard_offset
+        # How many list items are there in total?
+        n_items = self.load_worldsize * shard_len
+        # The indices of the flattened sharded_list that this worker owns
+        my_items = range(
+            int(n_items * self.rank / self.worldsize) - item_offset,
+            int(n_items * (self.rank + 1) / self.worldsize) - item_offset,
+        )
+        # Pull out owned items
+        return [sharded_list[i // shard_len][i % shard_len] for i in my_items]
```

Comment:
Yes your update is correct. We never actually materialize the full list of doc indices owned by each worker, because as you say, this gets very large.